### PR TITLE
Add architecture tests enforcing agent package boundaries

### PR DESCRIPTION
**Pull request by khaong**
## Summary
- Adds CI-enforced architecture tests that prevent agent implementations from coupling to framework internals (strategy, checkpoint, session, commands, top-level cli package)
- Relaxes the researcher skill's "Do NOT read source files" instruction — now that the boundary is enforced by tests, the researcher can reference `agent.go` and `event.go` for exact interface signatures

## What the tests do
- **`TestAgentPackages_NoForbiddenImports`**: Scans all agent package imports and fails if they reference forbidden internal packages. New agent packages are auto-discovered.
- **`TestAgentPackages_SelfRegister`**: Verifies every agent package has an `init()` calling `Register()` for proper discovery.

## Why
From reviewing #555 — the researcher skill had a hard "Do NOT read internal source files" instruction to prevent agents from building on internals. But agent instructions are a weak enforcement mechanism. These tests enforce the real boundary at the code level, so the doc instruction can be relaxed to useful guidance instead.

## Test plan
- [x] `mise run fmt && mise run lint` — clean
- [x] Architecture tests pass for all 4 agents (claudecode, geminicli, opencode, cursor)
- [x] Tests auto-discover new agent packages — no maintenance needed when adding agents

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: cmd/entire/cli/agent/architecture_test.go [hunk 1]
```diff
@@ -0,0 +1,247 @@
...
+				allowed := false
+				for _, prefix := range allowedPrefixes {
+					if imp == prefix || strings.HasPrefix(imp, prefix+"/") {
+						allowed = true
+						break
+					}
+				}
+				if !allowed {
+					t.Errorf("unexpected internal import %q — if this is intentional, add it to allowedPrefixes in architecture_test.go", imp)
+				}
```
[the remark this answers is not in the document]
Comment (khaong) [reply cmd/entire/cli/agent/architecture_test.go:105]: Fixed in 0051efb — added a `isForbidden` flag and `continue` after the forbidden suffix loop so it no longer falls through to the allowlist check.

[the remark this answers is not in the document]
Comment (khaong) [reply cmd/entire/cli/agent/architecture_test.go:105]: Fixed in 0051efb — same fix as above, added `continue` after forbidden suffix detection.

File: cmd/entire/cli/agent/architecture_test.go [hunk 2]
```diff
@@ -0,0 +1,247 @@
...
+	for _, pkgDir := range agentPkgs {
+		pkgName := filepath.Base(pkgDir)
+		t.Run(pkgName, func(t *testing.T) {
+			t.Parallel()
+			if !hasInitWithRegister(t, pkgDir) {
+				t.Errorf("agent package %q has no init() function calling Register() — agents must self-register", pkgName)
+			}
+		})
+	}
+}
```
[the remark this answers is not in the document]
Comment (khaong) [reply cmd/entire/cli/agent/architecture_test.go:211]: Fixed in 0051efb — added the same zero-package guard that `NoForbiddenImports` has.

File: cmd/entire/cli/agent/architecture_test.go [hunk 3]
```diff
@@ -0,0 +1,247 @@
...
+func findAgentDir(t *testing.T) string {
+	t.Helper()
+	// Walk up from the test file location to find the repo root.
+	wd, err := os.Getwd()
+	if err != nil {
+		t.Fatalf("os.Getwd: %v", err)
+	}
+	// We're in cmd/entire/cli/agent/ since the test is in that package.
+	return wd
+}
```
[the remark this answers is not in the document]
Comment (khaong) [reply cmd/entire/cli/agent/architecture_test.go:121]: Fixed in 0051efb — updated the comment. The implementation itself is correct since Go's test runner sets cwd to the package directory, so `os.Getwd()` reliably gives us the agent dir.

File: cmd/entire/cli/agent/architecture_test.go [hunk 4]
```diff
@@ -0,0 +1,247 @@
...
+
+	// Individual files in the cli package that agents must not depend on.
+	// We check for the cli package import itself since these are all in
+	// the top-level cli package (not sub-packages).
+	forbiddenCLIPackage := repoPrefix[:len(repoPrefix)-1] // "github.com/entireio/cli/cmd/entire/cli"
+
+	// Packages within the repo that agents ARE allowed to import.
+	allowedPrefixes := []string{
+		repoPrefix + "agent",      // agent contract + sub-packages
+		repoPrefix + "logging",    // logging utilities
```
[the remark this answers is not in the document]
Comment (khaong) [reply cmd/entire/cli/agent/architecture_test.go:47]: No agent currently cross-imports another, and the forbidden suffixes already cover the dangerous internals (strategy, checkpoint, session, commands). Adding a rule for agent-to-agent imports would be over-engineering for a scenario that doesn't exist. If it becomes a real concern we can add it then.

File: cmd/entire/cli/agent/architecture_test.go [hunk 5]
```diff
@@ -0,0 +1,247 @@
...
+					switch fun := call.Fun.(type) {
+					case *ast.Ident:
+						if fun.Name == "Register" {
+							found = true
+						}
+					case *ast.SelectorExpr:
+						if fun.Sel.Name == "Register" {
+							found = true
+						}
+					}
```
[the remark this answers is not in the document]
Comment (khaong) [reply cmd/entire/cli/agent/architecture_test.go:250]: In practice the only `Register` function agent packages call in `init()` is `agent.Register` — no other imported package has a `Register` function used in init. The unqualified `Register` case is also correct since the test file is in `package agent`. Tightening this would add complexity for a false positive that can't realistically occur.

**Comment by khaong** [comment]
bugbot run

File: cmd/entire/cli/agent/architecture_test.go [hunk 6]
```diff
@@ -0,0 +1,256 @@
...
+	if err != nil {
+		t.Fatalf("parser.ParseDir(%s): %v", dir, err)
+	}
+
+	seen := make(map[string]bool)
+	for _, pkg := range pkgs {
+		for _, file := range pkg.Files {
+			for _, imp := range file.Imports {
+				path := strings.Trim(imp.Path.Value, `"`)
+				seen[path] = true
```
[the remark this answers is not in the document]
Comment (khaong) [reply cmd/entire/cli/agent/architecture_test.go:177]: Fixed in 9355394 — switched to `strconv.Unquote` which handles both interpreted and raw string literals correctly.

**Review by Soph — APPROVED** [review]
